framework/frequency_manager.hpp

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2022 Intel Corporation.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef FREQUENCY_MANAGER_HPP
+#define FREQUENCY_MANAGER_HPP
+
+#define BASE_FREQ_PATH         "/sys/devices/system/cpu/cpu"
+#define SCALING_GOVERNOR       "/cpufreq/scaling_governor"
+#define SCALING_SETSPEED       "/cpufreq/scaling_setspeed"
+
+class FrequencyManager
+{
+private:
+#ifdef __linux__
+    int max_frequency_supported;
+    int min_frequency_supported;
+    std::vector<std::string> per_cpu_initial_scaling_governor;
+    std::vector<std::string> per_cpu_initial_scaling_setspeed;
+    int current_set_frequency;
+    std::vector<int> frequency_levels;
+    int frequency_level_idx = 0;
+    static constexpr int total_frequency_levels = 9;
+#endif
+
+    std::string read_file(const std::string &file_path)
+    {
+#ifdef __linux__
+        /* Read first line of given file */
+        char line[100]; //100 characters should be more than enough
+        FILE *file = fopen(file_path.c_str(), "r");
+
+        if (file == NULL) {
+            fprintf(stderr, "%s: cannot read from file: %s :%m\n", program_invocation_name, file_path.c_str());
+            exit(EXIT_FAILURE);
+        }
+        fscanf(file, "%s", line);
+        fclose(file);
+        return std::string(line);
+#else
+        return "";
+#endif
+    }
+
+    int write_file(const std::string &file_path, const std::string &line)
+    {
+#ifdef __linux__
+        FILE *file = fopen(file_path.c_str(), "w");
+
+        if (file == NULL)
+            return 1;
+
+        fprintf(file, "%s", line.c_str());
+        fclose(file);
+#endif
+        return 0;
```

Comment:
Please change this non-standard return code. Choose one of:
1. `bool` return for `true` for success and `false` for failure
2. `int` return with 0 for success and -1 for failure (like most POSIX functions)
3. `int` return with 0 for success and an errno number for failure

I prefer the first. The last one is unnecessary because we have `errno` .

Alternatively, you can make this function never fail and do `exit` if it would have. That would avoid duplication of the `exit(EXIT_NOPERMISSION)` code below.

Reply:
Done. I used the latter.